scripts/download-webkit.sh

```diff
@@ -43,11 +43,34 @@ fi
 
 rm -rf "$OUTDIR"
 
+download () {
+  local command="$1"
+  local retries="$2"
+  local options="$-"
+  if [[ $options == *e* ]]; then
+    set +e
+  fi
+  $command
+  local exit_code=$?
+  if [[ $options == *e* ]]; then
+    set -e
+  fi
+  if [[ $exit_code -ne 0 && $retries -gt 0 ]]; then
+    download "$command" $(($retries - 1)) 
+  else
+    return $exit_code
+  fi
+}
+
+# this is a big download so we will retry 5 times and ask curl to resume
+# download from where failure occurred if it fails and is rerun
 if [ ! -f "$tar" ]; then
   echo "-- Downloading WebKit"
-  if ! curl -o "$tar" -L "$url"; then
+  if ! download "curl -C - --http1.1 -o $tar.tmp -L $url" 5; then
```

Comment:
This is nice, but I wonder if we could use the `--retry` option in curl instead of doing this manually?

```sh
curl --retry 5 --retry-all-errors $tar.tmp -L $url
```

Reply:
lol. yes. i'll change it to use that.

Reply:
tbh I didn't know about it before googling

Reply:
will have to test it out to see if it works when combined with the resuming downloads flag. this suggests it will start from scratch on each retry.
https://github.com/curl/curl/issues/1084

Reply:
i did a few experiments with curl using tcpkill to kill the tcp connection for an active download. this is what i found, and it seems to agree with what the manpages for curl say. upshot i think is the method in this PR is only way to resume partial downloads from where they left off reliably.

1. ```--retry N``` only deals with http status errors, not tcp errors for an in-flight download
2. ```--retry-all-errors``` will always throw away existing partial download when it gets a tcp/network error mid-download (log below)

```shell
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  7 57.8M    7 4455k    0     0   305k      0  0:03:13  0:00:14  0:02:59  288k
curl: (56) OpenSSL SSL_read: Connection reset by peer, errno 104
Warning: Problem (retrying all errors). Will retry in 4 seconds. 3 retries 
Warning: left.
Throwing away 4562597 bytes
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  4 57.8M    4 2701k    0     0   557k      0  0:01:46  0:00:04  0:01:42  659k
```

Reply:
the curl manpages actually say this re. ```--retry-all-errors```:

https://curl.se/docs/manpage.html


This option is the "sledgehammer" of retrying. Do not use this option by default (for example in your curlrc), there may be unintended consequences such as sending or receiving duplicate data. Do not use with redirected input or output. **You might be better off handling your unique problems in a shell script**. Please read the example below.